API Doc for RadioButtons (#7)
Microsoft/microsoft-ui-xaml-specs: active/radiobuttons/RadioButtons_Spec.md

|MaximumColumns = 3|
|:--|
|![alt text](multicolumn_example_2.png)|

```xml
<RadioButtons Header="Select Number" MaximumColumns="3">
    <x:String>1</x:String>
    <x:String>2</x:String>
    <x:String>3</x:String>
    <x:String>4</x:String>
    <x:String>5</x:String>
    <x:String>6</x:String>
    <x:String>7</x:String>
</RadioButtons>
```

#### Keyboard navigation with multiple columns
The keyboarding behavior is the same as the single-column navigation, it just wraps to the next column when there is more than one defined.

![alt text](keyboardnav_multicol.png)

## Selection follows Focus
When you are navigating a RadioButtons list via the keyboard, as focus is placed on a previous or next RadioButton item, that item will also get selected/checked. This means the previous item that was selected/checked will be unselected, and the currently focused one will be instead.

|Before keyboard navigation | After keyboard navigation|
|:--|:--|
| ![alt text](2selected_nonav.png) | ![alt text](3selected_yesnav.png)|

> | Focus is on the "2" RadioButton, and it is shown as selected | The down or right arrow key has been pressed, so focus was moved to the "3" RadioButton, thus selected 3 and unselected 2. |

Reviewer: >right [](start = 77, length = 5)

In this single column case, pressing right should not change focus, correct?

Author: I believe that that is what we settled on in the last meeting @chigy to confirm.

Once I get confirmation, I'll edit this API.

Reviewer 2: > I believe that that is what we settled on in the last meeting @chigy to confirm.
> 
> Once I get confirmation, I'll edit this API.

Based on the old thread, this is what we settled last so sounds right to me.
-	Have up arrow move previous and down arrow move next
-	Left and right arrow keys move spatially

# Background

> Radio buttons are used for mutually exclusive options in an app. When the users selects

Reviewer 3: ```suggestion
Radio buttons are used for mutually exclusive options in an app. When the user selects
```

> one of a group, the previously-checked button is unchecked. Consequently, radio buttons

Reviewer 3: ```suggestion
one in a group, the previously-checked button is unchecked. Consequently, radio buttons
```

must be grouped, which is done with the Xaml [RadioButton](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.RadioButton)
control by setting the GroupName property on each button to the same value.

That mechanism is awkward, and doesn't by default give the right keyboarding or
accessibility behavior.

The RadioButtons control in this spec is introduced to help solve those issues.
It's like a ListView control, but the items in the list are RadioButton controls.

# Description

RadioButtons is a new control that enables you to create groups of RadioButton elements easily,

> while also correctly supporting keyboarding and narrator functionality.

Reviewer 3: ```suggestion
while also correctly supporting keyboarding and Narrator functionality.
```

# Examples

> The following example demonstrates how to create a small collective group of RadioButton elements

Reviewer 3: ```suggestion
The following example demonstrates how to create a group of RadioButton elements
```

using the RadioButtons control.

```xaml
<muxc:RadioButtons Header="App Mode" SelectedIndex="2">
    <x:String>Item 1</x:String>
    <x:String>Item 2</x:String>
    <x:String>Item 3</x:String>         
</muxc:RadioButtons>
```

![alt text](DefaultRadioButtonGroup.png)

## DataBinding with RadioButtons
An example of simple data binding of the ItemsSource with RadioButtons.

```xaml
<muxc:RadioButtons Header="App Mode" ItemsSource="{x:Bind radioButtonItems}" />
```

```C#
public sealed partial class MainPage : Page
{
    public class OptionDataModel
    {
        public string Label;
        public override string ToString()
        {
            return Label;
        }
    }

    List<OptionDataModel> radioButtonItems;

    public MainPage()
    {
        this.InitializeComponent();

        radioButtonItems = new List<OptionDataModel>();
        radioButtonItems.Add(new OptionDataModel() { Label = "Item 1" });
        radioButtonItems.Add(new OptionDataModel() { Label = "Item 2" });
        radioButtonItems.Add(new OptionDataModel() { Label = "Item 3" });
    }
}
```

## Vertical and horizontal RadioButton lists

> In instances where there is more that one column of RadioButton items within a single RadioButtons list control, the list will flow a column-major layout.

Reviewer 3: ```suggestion
In instances where there is more that one column of RadioButton items within a single RadioButtons list control, the list will flow in a column-major layout.
```

Maybe I'm the only one confused by this column-major wording... I find the "fill itself vertically" statement easier to understand.

Author: We define and use column-major in the docs [here](https://docs.microsoft.com/en-us/windows/uwp/design/input/keyboard-interactions#wrapping-homogeneous-list-and-grid-view-items) - so I'm attempting to use the same terminology in this doc as well.

|MaxColumns not specified <br>OR<br> MaxColumns = 1 |  MaxColumns = 4 |
|:--:|:--:|
| ![alt text](singlecolumn_example.png) | ![alt text](multicolumn_example_1.png) |

```xml

> <RadioButtons Header="Select Number" MaxColumns="2">

Reviewer 3: Does this belong in the next section, since it specifies 2 columns for 4 items?

    <x:String>1</x:String>
    <x:String>2</x:String>
    <x:String>3</x:String>
    <x:String>4</x:String>
</RadioButtons>
```

By default, the RadioButtons list will fill itself vertically, meaning if no ``MaxColumns`` value is set, it will be assumed to be 1 and orient itself vertically.

### Multiple columns (uncommon case)

> If you specify a ``MaxColumns`` value that is not equal to the number of items in the RadioButtons list, the list will arrange  itself in column-major order and put any uneven/remaining items in the first column, in the case of an uneven ``MaxColumns`` value or list items defined.

Reviewer 3: I was following along but got lost in this sentence by the time I got to "in the case of an uneven ``MaxColumns`` value or list items defined." 😅

|MaxColumns = 3|
|:--|
|![alt text](multicolumn_example_2.png)|

```xml
<RadioButtons Header="Select Number" MaxColumns="3">
    <x:String>1</x:String>
    <x:String>2</x:String>
    <x:String>3</x:String>
    <x:String>4</x:String>
    <x:String>5</x:String>
    <x:String>6</x:String>
    <x:String>7</x:String>
</RadioButtons>
```

# Remarks

> The RadioButtons control has special navigation behavior that helps not only keyboard accelerant users, but also accessibility users to navigate the list more quickly and more easily.

Reviewer 3: ```suggestion
The RadioButtons control has special navigation behavior that helps keyboard users navigate the list more quickly and more easily.
```

## Putting focus into the RadioButtons list

> There are two states that the RadioButtons control can have with it's RadioButton content:

Reviewer 3: ```suggestion
There are two states that the RadioButtons control can have with its RadioButton content:
```

### No RadioButton is selected
When no RadioButton controls are selected in the RadioButtons list, focus is put on the first RadioButton in the list.

> Note: The RadioButton that receives tab focus from this initial navigation will **not be selected/checked**.

|Focus is on button<br>RadioButton group does not have focus | Tab has been pressed<br>RadioButton group now has focus|
|:--:|:--:|
| ![alt text](noselecteditem_notabfocus.png) | ![alt text](noselecteditem_tabfocus.png)|

## Keyboard navigation
When you have a single column list of RadioButton control options and you have already put focus onto an item, you can navigate the RadioButtons list in logical sequential order using the keyboard.

- The up or down arrow keys will move to the "next" or "previous" logical item (defined in markup)
- The left or right arrow keys will move spatially

![alt text](keyboardnav_singlecol.png)

This means that in purely vertical RadioButton group lists, left/right arrow keys will not do anything, but up/down will still navigate as expected. However, in purely horizontal RadioButton group lists, left and right will navigate the same way as up and down.

> >This behavior is defined more below when we discuss multi-column RadioButton group lists.

Reviewer 3: This note doesn't really feel necessary to me.

### Keyboard navigation with multiple columns
The keyboarding behavior is the same as the single-column navigation, it just wraps to the next column when there is more than one defined.

![alt text](keyboardnav_multicol.png)

## Selection follows focus

> When you are navigating a RadioButtons list via the keyboard, as focus is placed on a previous or next RadioButton item, that item will also get selected/checked. This means the previous item that was selected/checked will be unselected, and the currently focused one will be instead.

Reviewer 3: ```suggestion
When you are navigating a RadioButtons list via the keyboard, as focus is placed on a previous or next RadioButton item, that item will also get selected/checked. This means the previous item that was selected/checked will be unselected.
```

|Before keyboard navigation | After keyboard navigation|
|:--|:--|
| ![alt text](2selected_nonav.png) | ![alt text](3selected_yesnav.png)|

> | Focus is on the "2" RadioButton, and it is shown as selected | The down or right arrow key has been pressed, so focus was moved to the "3" RadioButton, thus selected 3 and unselected 2. |

Reviewer 3: ```suggestion
| Focus is on the "2" RadioButton, and it is shown as selected | The down arrow key has been pressed, so focus was moved to the "3" RadioButton, thus selected 3 and unselected 2. |
```

Right arrow in this example will do nothing, correct?

> ### No Selection follows focus with gamepad

Reviewer 3: ```suggestion
### Navigating with gamepad
```

If you're navigating the RadioButtons container with a gamepad controller, the selection follows focus behavior will be disabled. When navigating a RadioButtons control, no gamepad engagement is necessary to initiate navigation or to select RadioButton elements within the group.

## Accessibility behavior

Below is a detailed table of, depending on the user interaction, what the narrator is expected to say.

|Upon *first* putting focus into the group | Navigating to a selected item |
|:--|:--|
| "Group name" RadioButton collection, x of N selected | RadioButton "name" selected, x of N |

> |"Group name" RadioButton collection control, none selected| RadioButton "name" not selected, x of N <br> *(If navigating with shift-arrow keys, meaning no selection following focus)* |

Reviewer 3: Is the word "control" deliberately included in this example but not the one above it?

> ## No Wrapping

Reviewer 3: This is kind of an odd section title, don't you think? Maybe should be included as a subsection in a Keyboard Navigation section instead?

> The RadioButtons group does not wrap. This is because when using a screen reader, a sense of boundary and clear indication of ends versus beginnings is lost and thus making it difficult for vision accessible users to navigate the list easily. There is also no enumeration with a RadioButtons group, since groups like this are meant to have a resonable number of items within.

Reviewer 3: ```suggestion
The RadioButtons group does not wrap. This is because when using a screen reader, a sense of boundary and clear indication of ends versus beginnings is lost and thus making it difficult for blind users to navigate the list easily. There is also no enumeration with a RadioButtons group, since groups like this are meant to have a resonable number of items within.
```

I've heard that "blind" is the more appropriate term

> The RadioButtons group does not wrap. This is because when using a screen reader, a sense of boundary and clear indication of ends versus beginnings is lost and thus making it difficult for vision accessible users to navigate the list easily. There is also no enumeration with a RadioButtons group, since groups like this are meant to have a resonable number of items within.

Reviewer 3: What does the "no enumeration" comment mean? Is that not the "x of N" mentioned previously?

> If you find that you need wrapping behavior, perhaps a RadioButtons group isn't for you. Consider using a [ListView with multiple selection enabled](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/listview-and-gridview#item-selection-and-interaction).

Reviewer 3: Not sure I would give this advice... maybe just remove this entirely?

# API Details

## API Notes

| Name | Description |
|:-|:--|
| Header | Places a label object above the group of RadioButton elements and is exposed to the UIA tree via the DataContext property on the RadioButtons control's ContentPresenter. |
| SelectedItem | Gets the content of the selected RadioButton item. Selection is denoted by that item's RadioButton icon being checked. |
| SelectedIndex | Gets or sets the selected RadioButton item by index. Selection is denoted by that item's RadioButton icon being checked.|
| Items | Gets or sets an object source within the RadioButtons group. All items placed within the RadioButtons control will be in the content of a generated RadioButton. |
| ItemsSource | Gets or sets an object source used to generate the content of the ItemsControl. |
| ItemTemplate | The template set on any generated RadioButton elements within the RadioButtons control. If the root of the ItemTemplate is a RadioButton it will be used as the generated container. |
| MaxColumns | Defines the number of columns used to determine the control's layout |

> | ColumnMajorUniformToLargestGridLayout | To be defined. |

Reviewer 3: ?

Author: I need to sync with Stephen on what this does.

### A RadioButton is selected
When a RadioButton is selected and the user tabs into the list, focus is put on the selected RadioButton.

|Focus is on button<br>RadioButton group does not have focus| Tab has been pressed<br>RadioButton group now has focus|
|:--:|:--:|
| ![alt text](selecteditem_notabfocus.png) | ![alt text](selecteditem_tabfocus.png)|

### No RadioButton is selected
When no RadioButton controls are selected in the RadioButtons list, focus is put on the first RadioButton in the list.

|Focus is on button<br>RadioButton group does not have focus | Tab has been pressed<br>RadioButton group now has focus|
|:--:|:--:|
| ![alt text](noselecteditem_notabfocus.png) | ![alt text](noselecteditem_tabfocus.png)|

## Keyboard navigation
When you have a single column list of RadioButton control options and you have already put focus onto an item, you can navigate the RadioButtons list in logical sequential order using the keyboard.

- The up or down arrow keys will move to the "next" or "previous" logical item (defined in markup)
- The left or right arrow keys will move spatially

![alt text](keyboardnav_singlecol.png)

> This means that in purely vertical RadioButton group lists, left/right arrow keys will not do anything, but up/down will still navigate as expected. However, in purely horizontal RadioButton group lists, left and right will navigate the same way as up and down.

Reviewer 4: What's the goal of this restriction? 
I tried this out and it feels weird to me that up-down navigates for a purely horizontal group but left-right does nothing for a purely vertical group. It's logically consistent as an implementation detail but as an end-user the behavior difference feels a bit weird.

# Background
Radio buttons are used for mutually exclusive options in an app. When the users selects
one of a group, the previously-checked button is unchecked. Consequently, radio buttons
must be grouped, which is done with the Xaml [RadioButton](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.RadioButton)
control by setting the GroupName property on each button to the same value.

That mechanism is awkward, and doesn't by default give the right keyboarding or
accessibility behavior.

The RadioButtons control in this spec is introduced to help solve those issues.
It's like a ListView control, but the items in the list are RadioButton controls.

# Description

RadioButtons is a new control that enables you to create groups of RadioButton elements easily,

> while also correctly supporting keyboarding and narrator functionality.

Reviewer 5: I'm sure we don't have any ties specifically to Narrator in the implementation, so we shouldn't say so here. (Talk about accessibility in general, not Narrator in particular.)

> # Examples

Author: I believe so, @StephenLPeters that should be possible, correct? If so I'll go ahead and add an example.

> # Examples

Reviewer 5: Sorry, my question has a typo and I'm not sure what the original question was.

> Note: The RadioButton that receives tab focus from this initial navigation will **not be selected/checked**.

|Focus is on button<br>RadioButton group does not have focus | Tab has been pressed<br>RadioButton group now has focus|
|:--:|:--:|
| ![alt text](noselecteditem_notabfocus.png) | ![alt text](noselecteditem_tabfocus.png)|

## Keyboard navigation
When you have a single column list of RadioButton control options and you have already put focus onto an item, you can navigate the RadioButtons list in logical sequential order using the keyboard.

- The up or down arrow keys will move to the "next" or "previous" logical item (defined in markup)
- The left or right arrow keys will move spatially

![alt text](keyboardnav_singlecol.png)

This means that in purely vertical RadioButton group lists, left/right arrow keys will not do anything, but up/down will still navigate as expected. However, in purely horizontal RadioButton group lists, left and right will navigate the same way as up and down.

>This behavior is defined more below when we discuss multi-column RadioButton group lists.

> ## Vertical and horizontal RadioButton lists

Author: Fixed!

|:--|
|![alt text](multicolumn_example_2.png)|

```xml
<RadioButtons Header="Select Number" MaximumColumns="3">
    <x:String>1</x:String>
    <x:String>2</x:String>
    <x:String>3</x:String>
    <x:String>4</x:String>
    <x:String>5</x:String>
    <x:String>6</x:String>
    <x:String>7</x:String>
</RadioButtons>
```

#### Keyboard navigation with multiple columns
The keyboarding behavior is the same as the single-column navigation, it just wraps to the next column when there is more than one defined.

![alt text](keyboardnav_multicol.png)

## Selection follows Focus
When you are navigating a RadioButtons list via the keyboard, as focus is placed on a previous or next RadioButton item, that item will also get selected/checked. This means the previous item that was selected/checked will be unselected, and the currently focused one will be instead.

|Before keyboard navigation | After keyboard navigation|
|:--|:--|
| ![alt text](2selected_nonav.png) | ![alt text](3selected_yesnav.png)|
| Focus is on the "2" RadioButton, and it is shown as selected | The down or right arrow key has been pressed, so focus was moved to the "3" RadioButton, thus selected 3 and unselected 2. |

> ## Narrator behavior

Author: Renamed section. The keyboard covers ideal keyboarding accessibility behavior, the only other outstanding behavior is narrator/screen reader.

If you find that you need wrapping behavior, perhaps a RadioButtons group isn't for you. Consider using a [ListView with multiple selection enabled](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/listview-and-gridview#item-selection-and-interaction).

# API Details

## RadioButtons IDL

``` c++
namespace MU_XC_NAMESPACE
{

[WUXC_VERSION_PREVIEW]
[webhosthidden]
[contentproperty("Items")]
[MUX_PROPERTY_CHANGED_CALLBACK(TRUE)]
[MUX_PROPERTY_CHANGED_CALLBACK_METHODNAME("OnPropertyChanged")]
unsealed runtimeclass RadioButtons : Windows.UI.Xaml.Controls.Control
{
    RadioButtons();

    Object ItemsSource;
    Windows.Foundation.Collections.IVector<Object> Items{ get; };
    [MUX_DEFAULT_VALUE("winrt::RadioButtonsElementFactory{}")]
    Object ItemTemplate;

    Windows.UI.Xaml.UIElement ContainerFromIndex(Int32 index);

>     [MUX_DEFAULT_VALUE("-1")]

Author: Removed.

    Int32 SelectedIndex;
    Object SelectedItem{ get; };
    event Windows.UI.Xaml.Controls.SelectionChangedEventHandler SelectionChanged;

    [MUX_DEFAULT_VALUE("1")]

>     Int32 MaximumColumns;

Author: I changed all references from `MaximumColumns` to `MaxColumns` in this API - but please note this does not reflect the implementation yet. @StephenLPeters  please change as necessary. Thanks!

    static Windows.UI.Xaml.DependencyProperty ItemsSourceProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty ItemsProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty ItemTemplateProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty SelectedIndexProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty SelectedItemProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty MaximumColumnsProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty HeaderProperty{ get; };
}

}

```

## RadioButtons Primitives

``` c++
namespace MU_XCP_NAMESPACE
{
[WUXC_VERSION_PREVIEW]
[webhosthidden]
[default_interface]
unsealed runtimeclass RadioButtonsElementFactory : MU_XC_NAMESPACE.ElementFactory
{
    RadioButtonsElementFactory();
}

[WUXC_VERSION_PREVIEW]
[webhosthidden]
[default_interface]
unsealed runtimeclass ColumnMajorUniformToLargestGridLayout : MU_XC_NAMESPACE.NonVirtualizingLayout
{
    ColumnMajorUniformToLargestGridLayout();

    [MUX_PROPERTY_CHANGED_CALLBACK(TRUE)]
    Int32 MaximumColumns{ get; set; };
    static Windows.UI.Xaml.DependencyProperty MaximumColumnsProperty{ get; };

    [MUX_PROPERTY_CHANGED_CALLBACK(TRUE)]

>     Int32 ColumnSpacing{ get; set; };

Reviewer 6: Also should the type be Int or Float?

>     Int32 ColumnSpacing{ get; set; };

Reviewer 5: Should be a double like elsewhere (like [Grid.ColumnSpacing](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Grid.ColumnSpacing)).

    static Windows.UI.Xaml.DependencyProperty ColumnSpacingProperty{ get; };

    [MUX_PROPERTY_CHANGED_CALLBACK(TRUE)]
    Int32 RowSpacing{ get; set; };
    static Windows.UI.Xaml.DependencyProperty RowSpacingProperty{ get; };
}

}

```

> ## RadioButtons.xaml

Author: renamed.

``` xml
<Style TargetType="local:RadioButtons" BasedOn="{StaticResource DefaultRadioButtonsStyle}" />

<Style x:Key="DefaultRadioButtonsStyle" TargetType="local:RadioButtons">
    <Setter Property="IsTabStop" Value="False" />
    <Setter Property="TabNavigation" Value="Once" />
    <Setter Property="Template">
        <Setter.Value>
            <ControlTemplate TargetType="local:RadioButtons">
                <StackPanel>
                    <ContentPresenter Content="{TemplateBinding Header}"/>
                    <local:ItemsRepeater x:Name="InnerRepeater" ItemTemplate="{TemplateBinding ItemTemplate}">
                        <local:ItemsRepeater.Layout>
                            <primitives:ColumnMajorUniformToLargestGridLayout ColumnSpacing="7" RowSpacing="3"/>
                        </local:ItemsRepeater.Layout>
                    </local:ItemsRepeater>
                </StackPanel>
                </ControlTemplate>
        </Setter.Value>
    </Setter>
</Style>
```

> # API Notes

Author: Move to top of API Details section.

| Name | Description |
|:-:|:--|

> | Header | Places a label object above the group of RadioButton elements and is exposed to the UIA tree via the DataContext property on the RadioButtons control's ContentPresenter. |

Author: @StephenLPeters can you confirm whether or not RadioButtons is still doing this in your new implementation? If not, I can remove/update this.

| SelectedItem | Gets the content of the selected RadioButton item. Selection is denoted by that item's RadioButton icon being checked. |
| SelectedIndex | Gets or sets the selected RadioButton item by index. Selection is denoted by that item's RadioButton icon being checked.|

> | Items | Gets or sets an object source used to generate the content of the ItemsControl. All items placed within the RadioButtons control will get the content of a generated RadioButton. |

Reviewer 5: This description sounds like ItemsSource, not Items.

Author: Attempted to redefine it. Please double check for clarity @StephenLPeters 😄

# API Details

## RadioButtons IDL

``` c++
namespace MU_XC_NAMESPACE
{

[WUXC_VERSION_PREVIEW]
[webhosthidden]
[contentproperty("Items")]
[MUX_PROPERTY_CHANGED_CALLBACK(TRUE)]
[MUX_PROPERTY_CHANGED_CALLBACK_METHODNAME("OnPropertyChanged")]
unsealed runtimeclass RadioButtons : Windows.UI.Xaml.Controls.Control
{
    RadioButtons();

    Object ItemsSource;
    Windows.Foundation.Collections.IVector<Object> Items{ get; };
    [MUX_DEFAULT_VALUE("winrt::RadioButtonsElementFactory{}")]
    Object ItemTemplate;

    Windows.UI.Xaml.UIElement ContainerFromIndex(Int32 index);

    [MUX_DEFAULT_VALUE("-1")]
    Int32 SelectedIndex;

>     Object SelectedItem{ get; };

Reviewer 6: Why is this get-only?

Reviewer: ItemsRepeater does not  have a way to get a container from a data object. This is by design as a data object might be duplicated in your data set and then repeater would not know which items to return.   Technically we can do this by either walking the tree looking for a radiobutton who's content matches the supplied item, or by accessing private repeater headers, But in general I think collection controls should not expose selection by data item. given that repeater does not support  it.

Reviewer 6: I think that's a feature gap for Repeater that we should fill. It's ambiguous for ItemsControl too but it's still offered. Binding to SelectedItem is a very popular ask, I don't think we should take that feature away.

Reviewer 7: @jevansaks @StephenLPeters Absolutely! For proper MVVM use cases, being able to bind to the SelectedItem property I feel would be very much appreciated by us customers. At least I would definitely use it much rather than using the SelectedIndex property. And as you have pointed out, history (Treeview SelectedItem property) shows us that there is a good chance an issue will be opened if the RadioButtons control won't ship with a settable (and bindable) SelectedItem property 😉

    event Windows.UI.Xaml.Controls.SelectionChangedEventHandler SelectionChanged;

    [MUX_DEFAULT_VALUE("1")]
    Int32 MaximumColumns;
    Object Header;

    static Windows.UI.Xaml.DependencyProperty ItemsSourceProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty ItemsProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty ItemTemplateProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty SelectedIndexProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty SelectedItemProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty MaximumColumnsProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty HeaderProperty{ get; };
}

}

```

## RadioButtons Primitives

``` c++
namespace MU_XCP_NAMESPACE
{
[WUXC_VERSION_PREVIEW]
[webhosthidden]
[default_interface]

> unsealed runtimeclass RadioButtonsElementFactory : MU_XC_NAMESPACE.ElementFactory

Reviewer 6: @StephenLPeters does this type need to be public? If so, why? What do people do with it?

Reviewer: It is the default value for the ItemTemplate property on RadioButtons. I think we could get away with moving this to the private namespace.

Reviewer 6: But would it still be in the public API? You couldn't use it from XAML then, but RadioBUttons should be able to set it from its code-behind.

Reviewer: I think it needs to be a winrt type to be assignable to the dependency property?

{
    RadioButtonsElementFactory();
}

[WUXC_VERSION_PREVIEW]
[webhosthidden]
[default_interface]

> unsealed runtimeclass ColumnMajorUniformToLargestGridLayout : MU_XC_NAMESPACE.NonVirtualizingLayout

Reviewer 6: From the PR my understanding was this was internal, do we think it's necessary to be public? Can we just create and use it from code-behind for now?

Reviewer: In the PR I'm setting the repeater's layout in the RadioButtons template. We could move this to code behind and not expose this type, but we'd be preventing the consumer from setting their own layout by retemplating. (unless we can use the facny new ReadLocalValue api I just learned about to only set the layout if the consumer hasn't already set one).

Reviewer: The other aspect to this, which hasn't been updated in the xaml file cause I didn't give kiki an new one yet (sorry!) is that we are setting the layouts column and row spacing properties to theme resources in the controls resource dictionary. This would mean we'd have to do these theme resource look ups from code behind i think and that's been problematic in the past.

Reviewer 6: I see, then having it be public in primitives seems fine.
